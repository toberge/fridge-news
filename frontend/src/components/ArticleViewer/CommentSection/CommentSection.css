.comments {
  /*margin: 1%;*/
  margin-top: 2rem;
  padding: 2rem 2%;
  border-top: 0.2rem solid var(--section-text);
}

.comment {
  background-color: var(--section-background);
  padding: 1%;
  border-radius: 0.3rem;
  margin: 0.7rem 0;
}

.author {
  font-weight: bold;
}

/* TODO remove or keep and make generic? --> replace bootstrap? (fitting more in with my design) */

#comment-submit {
  background-color: var(--button-background);
  color: var(--button-text);
  border: 0.2rem;
  padding: 0.9rem 1.9rem;
  border-radius: 0.3rem;
  cursor: pointer;
  font-size: 1.2rem;
  -webkit-transition: all 0.25s ease;
  -moz-transition: all 0.25s ease;
  -ms-transition: all 0.25s ease;
  -o-transition: all 0.25s ease;
  transition: all 0.25s ease;
}

#comment-submit:focus,
#comment-submit:hover {
  background-color: var(--button-focus);
  outline: none;
}
