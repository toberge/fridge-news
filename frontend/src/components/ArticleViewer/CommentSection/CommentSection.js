// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import MarkdownRenderer from 'react-markdown-renderer';
import SimpleMDE from 'react-simplemde-editor';
import 'easymde/dist/easymde.min.css';
import { Button } from '../../widgets';

export default class CommentSection extends Component<{ articleID: number }> {
  comment: string = '';

  render() {
    return (
      <section className="comments">
        <h2>Comments</h2>
        <div className="comment">
          <p>
            <em>
              <span className="author">Grumbling Grevory</span> said:
            </em>
          </p>
          <p>I really hate this article...</p>
        </div>
        <div className="comment">
          <p>
            <em>
              <span className="author">Somebody</span> said:
            </em>
          </p>
          <p>
            Everything is fine.
            <br />
            It really is.
          </p>
        </div>
        <h2>Leave your own comment</h2>
        <form>
          <div className="form-group">
            <SimpleMDE onChange={this.handleMarkdownChange} label="Comment:" options={{ spellChecker: false }} />
          </div>
          <br />
          {/*<button id="comment-submit">Submit</button>*/}
          <Button.Primary onClick={() => null}>Submit</Button.Primary>
        </form>
      </section>
    );
  }

  handleMarkdownChange(value: string) {
    this.comment = value;
  }
}
