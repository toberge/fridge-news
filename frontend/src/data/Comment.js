// @flow

export default class Comment {
  articleID: number;
  authorID: number;
  text: string;
  uploadTime: Date;
  updateTime: Date | null;

  constructor(articleID: number, authorID: number, text: string, uploadTime: Date, updateTime: Date | null) {
    this.articleID = articleID;
    this.authorID = authorID;
    this.text = text;
    this.uploadTime = uploadTime;
    this.updateTime = updateTime;
  }
}
