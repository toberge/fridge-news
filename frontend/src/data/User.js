// @flow

export default class User {
  id: number;
  name: string;
  admin: boolean;

  constructor(id: number, name: string, admin: boolean) {
    this.id = id;
    this.name = name;
    this.admin = admin;
  }
}
