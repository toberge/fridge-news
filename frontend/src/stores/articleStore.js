// @flow

import axios from 'axios';
import { Article, ArticleBase } from '../data/Article';
import { sharedComponentData } from 'react-simplified';

const placeholder = new Article(
    1,
    1,
    'Fridge Found Floating in Space',
    'https://i.imgur.com/puQs66y.png',
    'Fridge floating in space',
    'The infamous space-floating fridge',
    `*After its long flight through the rings of Saturn, a fridge of mysterious
origin was caught by a spacewalking astronaut on the ISS and brought to
Earth for inspection. Inside were the remains of an unknown biological life
form - and a pile of ancient floppy disks. The floppies contained impressive
amounts of text - newspaper articles, essays, even a few poems. Here are a
few snippets from the fridge's collection.*
### Space Cows Demand We Go Vegan  
Yesterday, a fleet of extrasolar ships approached Earth, bearing an ultimatum.
If we do not stop eating meat and return all our livestock to nature, our entire civilization will be eradicated.
The species behind this harsh request look almost exactly like the cows of humanity's home world, but their intelligence is on a whole other level, perhaps exceeding ours.
Most governments in the solar system have already stated that they perceive this threat as very real, some adding that they will take immediate action to enforce a meat ban.`,
    'news',
    new Date(),
    null,
    1,
    2.3
  );

const empty = new Article(
    1,
    1,
    '',
    '',
    '',
    '',
    `*Begin your article with an ingress*
    
## Use headers of level 2 and below

write some **good** text`,
    '',
    new Date(),
    null,
    2,
    2.3
  );

class ArticleStore {
  // won't make loading multiple articles necessitate showing a loading page,
  // it's better to show a not-yet-updated list than to show *the wrong article*
  loadingArticle: boolean = false;
  currentArticle: Article = placeholder;
  articles: ArticleBase[] = [];
  categoryMap: Map<string, ArticleBase[]> = new Map<string, ArticleBase[]>([
    ['news', []],
    ['culture', []],
    ['science', []],
    ['politics', []]
  ]);

  categories: string[] = ['news', 'culture', 'science', 'politics', 'blablabla'];

  // wow shit why must I be forced to do this...
  getCategories() {
    return axios.get('/articles/categories/')
      .then(response => response.data)
      .then(strings => {
        // replace old array (if anything lives there)
        const countBefore = this.categories.length;
        this.categories.push(strings);
        this.categories.splice(0, countBefore);
        // add to hashmap of cached categories
        strings.forEach(s => {
          if (!this.categoryMap.get(s)) {
            this.categoryMap.set(s, []);
          }
        })
      })
  }

  clearArticle() {
    Object.assign(this.currentArticle, { ...empty });
  }

  getArticle(id: number) {
    this.loadingArticle = true;
    return axios
      .get<Article>('/articles/' + id)
      .then(response => response.data)
      .then(result => {
        const {
          article_id,
          user_id,
          title,
          picture_path,
          picture_alt,
          picture_caption,
          content,
          upload_time,
          update_time,
          importance,
          category,
          rating
        } = result;
        this.currentArticle = new Article(
          article_id,
          user_id,
          title,
          picture_path,
          picture_alt,
          picture_caption,
          content,
          category,
          new Date(Date.parse(upload_time)),
          update_time ? new Date(Date.parse(update_time)) : null,
          importance,
          rating
        );
        this.loadingArticle = false;
        return result;
      });
  }

  addArticle(): Promise<number | void> {
    const article = this.currentArticle;
    return axios.post('/articles/', {
      user_id: article.authorID,
      title: article.title,
      picture_path: article.picturePath,
      picture_alt: article.pictureAlt,
      picture_caption: article.pictureCapt,
      content: article.text,
      importance: article.importance,
      category: article.category
    }).then(response => response.data.id);
  }

  updateArticle(): Promise<boolean | void> {
    const article = this.currentArticle;
    return axios.put('/articles/' + article.id, {
      title: article.title,
      picture_path: article.picturePath,
      picture_alt: article.pictureAlt,
      picture_caption: article.pictureCapt,
      content: article.text,
      importance: article.importance,
      category: article.category
    }).then(response => response.status === 200);
  }

  deleteArticle(): Promise<boolean> {
    return axios.delete(`/articles/${this.currentArticle.id}`)
      .then(response => response.status === 200);
  }

  getFrontPage(): Promise<ArticleBase[]> {
    return axios
      .get<ArticleBase[]>('/articles/front_page')
      .then(response => {
        this.articles.splice(0, this.articles.length);
        this.articles.push(...this.toArticleArray(response.data));
      });
  }

  getCategory(category: string): Promise<ArticleBase[]> {
    return axios
      .get<ArticleBase[]>('/articles/categories/' + category)
      .then(response => {
        const array = this.categoryMap.get(category);
        if (array) {
          array.splice(0, array.length);
          array.push(...this.toArticleArray(response.data));
        }
      })
      /*.catch(e => {
        // this is temporary crap
        this.articles.splice(0, this.articles.length);
        throw e;
      })*/
  }

  toArticleArray(result: []): ArticleBase[] {
    return result.map(e => {
      const {article_id, title, picture_path, picture_alt, category} = e;
      return new ArticleBase(article_id, title, picture_path, picture_alt, category);
    });
  }
}

export const articleStore = sharedComponentData(new ArticleStore());
