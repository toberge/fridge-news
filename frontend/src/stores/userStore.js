// @flow

import axios, { AxiosRequestConfig, AxiosResponse } from 'axios';
import { sharedComponentData } from 'react-simplified';
import User from '../data/User';
import Notifier from '../components/shared/Notifier';

class UserStore {
  // login state
  currentUser: ?User;
  token: ?string;
  tokenInterval: ?IntervalID = null;
  // other users
  currentAuthor: User = new User(1, 'The Fridge', true);
  cachedUsers: Map<number, User> = new Map<number, User>();

  loadUser() {
    const username = sessionStorage.getItem('username');
    const id = sessionStorage.getItem('userID');
    const token = sessionStorage.getItem('token');
    if (username && id && token) {
      // set temp user
      this.currentUser = new User(parseInt(id), username, false);
      this.token = token;
      // load user from DB
      this.getUser(parseInt(id))
        .then(user => this.currentUser = user)
        .catch(error => {
          console.log(error);
          this.currentUser = null;
        });
    }
  }

  saveUser() {
    if (this.currentUser && this.token) {
      sessionStorage.setItem('username', this.currentUser.name);
      // excessive checks because Flow is stupid sometimes
      if (this.currentUser) sessionStorage.setItem('userID', `${this.currentUser.id}`);
      if (this.token) sessionStorage.setItem('token', this.token);
    }
  }

  logIn(name: string, password: string) {
    return axios.post('/login', { name, password }).then(async (response: AxiosResponse) => {
      // already okay
      console.log(response.data);
      this.currentUser = await this.getUser(response.data.user_id);
      this.token = response.data.jwt;
      this.startTokenInterval();
      this.saveUser();
      return true;
    });
  }

  /**
   * Just logs off our dear user
   */
  logOut() {
    this.currentUser = null;
    this.token = null;
    sessionStorage.clear();
    clearInterval(this.tokenInterval);
  }

  register(username: string, password: string) {
    return axios.post('/users/', { name: username, password }).then(async response => {
      this.currentUser = await this.getUser(response.data.insertId);
      this.token = response.data.jwt;
      this.startTokenInterval();
      this.saveUser();
      return this.currentUser ? this.currentUser.id : -1; // TODO temp hack...
    });
  }

  getTokenHeader(): AxiosRequestConfig {
    return {
      headers: {
        'x-access-token': userStore.token
      }
    };
  }

  startTokenInterval() {
    this.tokenInterval = setInterval(this.refreshToken, 1000 * 60 * 4);
  }

  refreshToken() {
    axios
      .get('/token/', this.getTokenHeader())
      .then((response: AxiosResponse) => {
        if (response.data.jwt) this.token = response.data.jwt;
        console.log('fetched token...');
      })
      .catch((e: Error) => {
        Notifier.error(`Refreshing token failed, logging out... Error: ${e.message}`);
        this.logOut();
      });
  }

  getUser(id: number): Promise<User> {
    let user = this.cachedUsers.get(id);
    if (user) {
      return new Promise<User>(resolve => resolve(user));
    } else {
      return axios.get('/users/' + id).then(response => {
        const { user_id, name, admin } = response.data;
        const newUser = new User(user_id, name, admin);
        this.cachedUsers.set(id, newUser);
        return newUser;
      });
    }
  }

  getAuthor(id: number) {
    return this.getUser(id).then(user => (this.currentAuthor = user));
  }
}

export const userStore = sharedComponentData(new UserStore());
