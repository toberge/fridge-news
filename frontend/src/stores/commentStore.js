// @flow

import axios from 'axios';
import { sharedComponentData } from 'react-simplified';
import Comment from '../data/Comment';
import {userStore} from "./userStore";

const empty = new Comment(-1, 1, '', new Date(), null);

class CommentStore {
  draft: Comment = empty;
  comments: Comment[] = [];

  /**
   * Fetches comments, filling userStore with authors if necessary.
   *
   * @param id
   * @returns {Promise<Comment[]>}
   */
  getComments(id: number): Promise<Comment[]> {
    return axios.get(`/articles/${id}/comments`)
      .then(response => {
        this.comments.splice(0, this.comments.length);
        this.comments.push(...this.toCommentArray(response.data));
      })
      .then(async data => {
        // fow complained about inside of forEach...
        // this.comments.forEach(c => (
        for (const c of this.comments) {
          await userStore.getUser(c.authorID);
        }
        // ));
        return data;
      })
  }

  resetDraft(id: number) {
    Object.assign(this.draft, {...empty});
    this.draft.articleID = id;
  }

  addComment(): Promise<*> {
    return axios.post(`/articles/${this.draft.articleID}/comments`, {
      user_id: this.draft.authorID, // NOT USED BTW
      content: this.draft.text
    })
  }

  toCommentArray(result: []): Comment[] {
    return result.map(e => {
      const { article_id, user_id, content, upload_time, update_time } = e;
      return new Comment(article_id, user_id, content, upload_time, update_time);
    });
  }
}

export const commentStore = sharedComponentData(new CommentStore());
