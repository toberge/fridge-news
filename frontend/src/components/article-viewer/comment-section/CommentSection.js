// @flow

import * as React from 'react';
import { Component } from 'react-simplified';
import MarkdownRenderer from 'react-markdown-renderer';
import SimpleMDE from 'react-simplemde-editor';
import 'easymde/dist/easymde.min.css';
import Form from '../../shared/Form';
import Icon from '../../shared/Icon';
import './CommentSection.css';
import { commentStore } from '../../../stores/commentStore';
import { userStore } from '../../../stores/userStore';
import Notifier from "../../shared/Notifier";

export default class CommentSection extends Component<{ articleID: number }> {
  comment: string = '';
  loading: boolean = false;
  pending: boolean = false;

  render() {
    return (
      <aside className="comments">
        <h2>Comments</h2>
        {this.loading ? (
          <em>Loading...</em>
        ) : commentStore.comments.length > 0 ? (
          commentStore.comments.map(c => (
            <div className="comment">
              <p>
                <em>
                  <span className="author">{this.getAuthor(c.authorID)}</span> said:
                </em>
              </p>
              <MarkdownRenderer markdown={c.text} />
            </div>
          ))
        ) : <em>No comments yet.</em>}
        <h2>Leave your own comment</h2>
        <form onSubmit={this.handleSubmit}>
          <div className="form-group">
            <SimpleMDE value={commentStore.draft.text} onChange={this.handleMarkdownChange} label="Comment:" options={{ spellChecker: false }} />
          </div>
          <Form.Submit disabled={this.pending || commentStore.draft.text === ''}><Icon.Upload /> Submit</Form.Submit>
        </form>
      </aside>
    );
  }

  handleMarkdownChange(value: string) {
    commentStore.draft.text = value;
  }

  getAuthor(id: number) {
    const user = userStore.cachedUsers.get(id);
    if (user) {
      return user.name;
    } else {
      return 'Unknown author';
    }
  }

  async mounted() {
    this.loading = true;
    commentStore.resetDraft(this.props.articleID);
    try {
      await commentStore.getComments(this.props.articleID);
      this.loading = false;
    } catch (e) {
      Notifier.error(`Could not fetch comments\n${e.message}`);
    }
  }

  async handleSubmit(event: SyntheticInputEvent<HTMLFormElement>) {
    event.preventDefault();

    if (commentStore.draft.text) {
      this.pending = true;
      try {
        await commentStore.addComment();
        Notifier.success('Uploaded comment!');
        await commentStore.getComments(this.props.articleID);
      } catch (e) {
        Notifier.error(`Could not upload comment\n${e.message}`);
      }
      this.pending = false;
    }
  }
}
