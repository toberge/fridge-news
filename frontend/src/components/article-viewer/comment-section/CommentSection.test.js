// @flow

import * as React from 'react';
import 'core-js/stable';
import 'regenerator-runtime/runtime';
import CommentSection from './CommentSection';
import Comment from '../../../data/Comment';
import { shallow, ShallowWrapper } from 'enzyme';
import { commentStore } from '../../../stores/commentStore';
import { userStore } from '../../../stores/userStore';
import User from '../../../data/User';

jest.mock('../../../stores/commentStore');

describe('CommentSection', () => {
  const commentSection: ShallowWrapper = shallow(<CommentSection articleID={1} />);

  it('should render no comments when none is present', () => {
    expect(commentSection.debug()).toMatchSnapshot();
  });

  it('should not be loading when getComments() is mocked', () => {
    expect(commentSection.instance().loading).toBe(false);
  });

  it('should fetch comments when mounting', () => {
    expect(commentStore.getComments).toHaveBeenCalled();
  });

  it('should list author as unknown if not found', () => {
    commentStore.comments.push(
      new Comment(1, 1, 'I hate this article', new Date(), null)
    );
    // supplying no user
    commentSection.update();
    commentSection.instance().forceUpdate();

    expect(commentSection.debug()).toMatchSnapshot();
  });

  it('should render comments when those are present', () => {
    commentStore.comments.push(
      new Comment(1, 1, "There, there... It'll be fine.", new Date(), null)
    );
    userStore.cachedUsers.set(1, new User(1, 'The Fridge', true));
    commentSection.update();
    commentSection.instance().forceUpdate();

    expect(commentSection.debug()).toMatchSnapshot();
  });



  it('should render form when user is logged in', () => {
    // log in
    userStore.currentUser = new User(1, 'The Fridge', true);
    commentSection.update();
    commentSection.instance().forceUpdate();

    expect(commentSection.debug()).toMatchSnapshot();
  });

  // it('should not upload w/o input', () => {
  it('should call handleSubmit() when submit is pressed', () => {
    // spy on dat method, mocking it to prevent the actual method from being called
    const submitSpy = jest.spyOn(commentSection.instance(), 'handleSubmit').mockImplementation(() => null);
    // log in
    userStore.currentUser = new User(1, 'The Fridge', true);
    commentSection.update();
    commentSection.instance().forceUpdate();

    //click and inspect
    expect(submitSpy).not.toHaveBeenCalled();
    commentSection.find('form').simulate('submit');
    expect(submitSpy).toHaveBeenCalled();
  });
});
