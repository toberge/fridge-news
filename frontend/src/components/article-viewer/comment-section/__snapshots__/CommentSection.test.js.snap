// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommentSection should render comments when those are present 1`] = `
"<aside className=\\"comments\\">
  <h2>
    Comments
  </h2>
  <div className=\\"comment\\">
    <p>
      <em>
        <span className=\\"author\\">
          Unknown author
        </span>
         said:
      </em>
    </p>
    <MarkdownRenderer markdown=\\"I hate this article\\" />
  </div>
  <div className=\\"comment\\">
    <p>
      <em>
        <span className=\\"author\\">
          Unknown author
        </span>
         said:
      </em>
    </p>
    <MarkdownRenderer markdown=\\"There, there... It\\\\'ll be fine.\\" />
  </div>
  <h2>
    Leave your own comment
  </h2>
  <p>
    You must 
    <Link to=\\"/login\\">
      log in
    </Link>
     to leave a comment
  </p>
</aside>"
`;

exports[`CommentSection should render form when user is logged in 1`] = `
"<aside className=\\"comments\\">
  <h2>
    Comments
  </h2>
  <div className=\\"comment\\">
    <p>
      <em>
        <span className=\\"author\\">
          Unknown author
        </span>
         said:
      </em>
    </p>
    <MarkdownRenderer markdown=\\"I hate this article\\" />
  </div>
  <div className=\\"comment\\">
    <p>
      <em>
        <span className=\\"author\\">
          Unknown author
        </span>
         said:
      </em>
    </p>
    <MarkdownRenderer markdown=\\"There, there... It\\\\'ll be fine.\\" />
  </div>
  <h2>
    Leave your own comment
  </h2>
  <form onSubmit={[Function: bound handleSubmit]}>
    <div className=\\"form-group\\">
      <SimpleMDEEditor value=\\"\\" onChange={[Function: bound handleMarkdownChange]} label=\\"Comment:\\" options={{...}} events={{...}} />
    </div>
    <FormSubmit disabled={true}>
      <UploadIcon />
       Submit
    </FormSubmit>
  </form>
</aside>"
`;

exports[`CommentSection should render no comments when none is present 1`] = `
"<aside className=\\"comments\\">
  <h2>
    Comments
  </h2>
  <em>
    No comments yet.
  </em>
  <h2>
    Leave your own comment
  </h2>
  <p>
    You must 
    <Link to=\\"/login\\">
      log in
    </Link>
     to leave a comment
  </p>
</aside>"
`;
